### PR TITLE
GEODE-3693: increase test coverage of CacheListener events

### PR DESCRIPTION
**Pull request by kirklund**
* add CacheEventListenerDUnitTest
* add PartitionedRegionEventsDUnitTest

Also:

GEODE-3747: add SharedErrorCollector rule
    
* use for assertions in callbacks and in multiple JVMs
* minor cleanup of related classes

### CODE REVIEW

File: geode-core/src/test/java/org/apache/geode/internal/cache/PartitionedRegionEventsDUnitTest.java [hunk 1]
```diff
@@ -0,0 +1,775 @@
...
+    getHost(0).getVM(0).invoke(
+        () -> updateNewEntry(name, new SubscriptionAttributes(InterestPolicy.ALL), listener));
+
+    assertThat(sharedCountersRule.getTotal(UPDATES)).isEqualTo(expectedUpdates);
+    assertThat(sharedCountersRule.getTotal(CREATES)).isEqualTo(expectedCreates);
+  }
+
+  @Test
+  public void testObjectInvalidated() throws Exception {
+    int timeToLive = 0; // 5; // seconds expiration isn't supported in PR (yet??)
```
Comment (dschneider-pivotal) [inline geode-core/src/test/java/org/apache/geode/internal/cache/PartitionedRegionEventsDUnitTest.java]: we do support expiration on prs. Seems very wrong to have a test that looks like it is testing expiration invalidates but actually just calls invalidate explicitly. If this test does not need to test expiration then we should remove the expiration logic from this test.

File: geode-core/src/test/java/org/apache/geode/internal/cache/PartitionedRegionEventsDUnitTest.java [hunk 2]
```diff
@@ -0,0 +1,775 @@
...
+
+    for (int i = 0; i < getHost(0).getVMCount(); i++) {
+      Object vmNewValue = getHost(0).getVM(i).invoke(() -> newValue);
+      assertThat(vmNewValue).as("newValue is wrong in vm " + i).isNull();
+    }
+  }
+
+  @Test
+  public void testObjectDestroyed() throws Exception {
+    int timeToLive = 0; // 3; expiration isn't supported by PRs (yet???)
```
Comment (dschneider-pivotal) [inline geode-core/src/test/java/org/apache/geode/internal/cache/PartitionedRegionEventsDUnitTest.java]: same goes for expiration in this test

File: geode-core/src/test/java/org/apache/geode/internal/cache/PartitionedRegionEventsDUnitTest.java [hunk 3]
```diff
@@ -0,0 +1,775 @@
...
+
+    for (int i = 0; i < getHost(0).getVMCount(); i++) {
+      Object id = getHost(0).getVM(i).invoke(() -> cacheRule.getSystem().getDistributedMember());
+      assertThat(bucketOwner).isNotEqualTo(id);
+    }
+  }
+
+  @Test
+  public void testObjectInvalidatedCACHECONTENT() throws Exception {
+    int timeToLive = 0; // 5; // seconds expiration isn't supported in PR (yet??)
```
Comment (dschneider-pivotal) [inline geode-core/src/test/java/org/apache/geode/internal/cache/PartitionedRegionEventsDUnitTest.java]: same for this test

File: geode-core/src/test/java/org/apache/geode/internal/cache/PartitionedRegionEventsDUnitTest.java [hunk 4]
```diff
@@ -0,0 +1,775 @@
...
+
+    for (int i = 0; i < getHost(0).getVMCount(); i++) {
+      Object id = getHost(0).getVM(i).invoke(() -> cacheRule.getSystem().getDistributedMember());
+      assertThat(bucketOwner).isNotEqualTo(id);
+    }
+  }
+
+  @Test
+  public void testObjectDestroyedCACHECONTENT() throws Exception {
+    int timeToLive = 0; // 3; expiration isn't supported by PRs (yet???)
```
Comment (dschneider-pivotal) [inline geode-core/src/test/java/org/apache/geode/internal/cache/PartitionedRegionEventsDUnitTest.java]: same for expiration in this test

**Comment by kirklund** [comment]
@dschneider-pivotal I separated GEODE-3747 out into its own PR: https://github.com/apache/geode/pull/867

I think we should hold off on committing these new tests (GEODE-3693) until I have time to rewrite these two tests as well as PartitionedRegionCloseDUnitTest and DeltaToRegionRelationDUnitTest. Those are the last 4 "legacy" dunit tests. Porting them to Geode and removing any flakiness they have is straightforward but rewriting them will probably take another week and a half.

**Pull request closed without merging**